text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1006 @@
+- Feature Name: variadic_tuples
+- Start Date: 2019-08-16
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+Tuples types are ordered set of type, but users can only use tuple with fixed arity.
+
+This RFC aims to allow the use of a _variadic tuple_ which is a tuple with an arbitrary arity.
+
+# Motivation
+
+[motivation]: #motivation
+
+## Arbitrary tuple arity support
+
+Currently, when a user wants to either use or add behavior to tuples, he writes an impl for each tuple arity.
+For easier maintenance, it is usually done with a `macro_rules` and implemented up to 12 arity tuple. (see [ `Hash` implementation in `std`](https://github.com/rust-lang/rust/blob/master/src/libcore/hash/mod.rs)).
+
+Variadic tuple will provide several benefits considering trait implementation for tuple or using variadic tuples:
+
+- Implementations will be easier to write
+- Implementations will be easier to read and maintain
+- The compiler will compile implementation only for required tuple arity
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Let's call a _variadic tuple type_ a tuple type with an arbitrary arity and a _variadic tuple_ an instance of a variadic tuple type.
```

Comment:
This is a very formal mathematical tone, and this section really could do without it to be a bit more approachable. First, you don't need to add a "let's call"; it's really more just, "A variadic tuple is…"

But really, another big thing is that the definition of "arity" is not clarified here, when the average user really needs to know why. It's not a common word that most programmers will know.

I'd start with something like…

> In Rust, you can combine any number of types together to form a tuple. For example, you could have a function which orders a meal that returns a `Drink`, a `Meal`, and a `Side`. We could represent this as the tuple `(Drink, Meal, Side)`.

> In this case, the tuple has three items, but really, we can have any number. If we have no items, we get `()`, the "empty tuple" or "unit" type. We can even put a type `T` in a tuple with only one item, whose type is `(T,)`.

> If we want to be able to cover *any* size of tuple, we use what's called a *variadic tuple type*, whose instances are *variadic tuples*.

Feel free to directly use that or modify it however you want.